modify verify.py to skip resource

tsutaj opened the pull request:
#235 の対応

処理自体はもう書けているつもりで、まだ手元の Windows でテストできてないんですが、GitHub Actions で通るかどうか確認するためにもとりあえず PR を建てます

(手元でも一応試したいので、それが確認できてから merge していただければ助かります)

tsutaj commented:
これ手順間違えていますね (cherry-pick を忘れている)
対応します

tsutaj commented:
最初で落ちなくなったけど謎に exit code 1 になっている・・・ (なんで？)

kmyk commented:
`import resource` をしないだけで、`resource.getrlimit` とかは実行されちゃうからですね。
使ってる場所全体を `if` で包んで、その `if` の中で `import resource` するのがよいと思います。

https://github.com/online-judge-tools/verification-helper/blob/fb6f59b77c6e697cba275c44be4ecee45bdecd9f/onlinejudge_verify/verify.py#L126-L131

tsutaj commented:
すみません、try except で囲まれているので問題ないかと誤解していました。修正します。

kmyk commented:
ごめんなさい、私の勘違いです。たしかに `try` `except` で囲まれてるのでセーフのはずですね (でも `if ...: import ...` をやるなら実際に使う場所に置いておいた方が分かりやすいので、修正自体はお願いしたいです)

kmyk commented:
普通に `-h` で help message を出力してるだけにしか見えないが、なぜ落ちるんだろう？ 分かりません

``` python
$ oj-verify -h
no module named resource (skipped)
usage: oj-verify [-h] [--config-file CONFIG_FILE] {all,run,docs} ...

positional arguments:
  {all,run,docs}

optional arguments:
  -h, --help            show this help message and exit
  --config-file CONFIG_FILE
                        default: ".verify-helper/config.toml"
##[error]Process completed with exit code 1.
```

tsutaj commented:
手元の素の Python で実行してみましたが、`/dev/stdout` 周りでこけているように見えます

```
(venv) C:\Users\tsutaj\tmp>oj-verify run examples\monoids.hpp
no module named resource (skipped)
INFO:onlinejudge_verify.config:config file loaded: .verify-helper\config.toml: {'languages': {'awk': {'compile': "bash -c 'echo hello > {tempdir}/hello'", 'execute': 'env AWKPATH={basedir} awk -f {path}', 'bundle': 'false', 'list_dependencies': 'sed \'s/^@include "\\(.*\\)"$/\\1/ ; t ; d\' {path}'}}}
WARNING:onlinejudge_verify.languages:config.toml: languages.awk: Adding new languages using `config.toml` is supported but not recommended. Please consider making pull requests for your languages, see https://github.com/kmyk/online-judge-verify-helper/issues/116
<built-in>: fatal error: opening dependency file /dev/stdout: No such file or directory
compilation terminated.
Traceback (most recent call last):
  File "C:\Users\tsutaj\tmp\venv\lib\site-packages\online_judge_verify_helper-4.10.1-py3.8.egg\onlinejudge_verify\marker.py", line 37, in get_current_timestamp
    depending_files = language.list_dependencies(path, basedir=pathlib.Path.cwd())
  File "C:\Users\tsutaj\tmp\venv\lib\site-packages\online_judge_verify_helper-4.10.1-py3.8.egg\onlinejudge_verify\languages\cplusplus.py", line 168, in list_dependencies
    return _cplusplus_list_depending_files(path.resolve(), CXX=env.CXX, joined_CXXFLAGS=joined_CXXFLAGS)
  File "C:\Users\tsutaj\tmp\venv\lib\site-packages\online_judge_verify_helper-4.10.1-py3.8.egg\onlinejudge_verify\languages\cplusplus.py", line 47, in _cplusplus_list_depending_files
    data = subprocess.check_output(command)
  File "C:\Users\tsutaj\AppData\Local\Programs\Python\Python38\lib\subprocess.py", line 411, in check_output
    return run(*popenargs, stdout=PIPE, timeout=timeout, check=True,
  File "C:\Users\tsutaj\AppData\Local\Programs\Python\Python38\lib\subprocess.py", line 512, in run
    raise CalledProcessError(retcode, process.args,
subprocess.CalledProcessError: Command '['C:\\MinGW\\bin\\g++.EXE', '--std=c++17', '-O2', '-Wall', '-g', '-I', 'C:\\Users\\tsutaj\\tmp', '-MD', '-MF', '/dev/stdout', '-MM', 'C:\\Users\\tsutaj\\tmp\\example.test.cpp']' returned non-zero exit status 1.
<built-in>: fatal error: opening dependency file /dev/stdout: No such file or directory
compilation terminated.
Traceback (most recent call last):
  File "C:\Users\tsutaj\tmp\venv\lib\site-packages\online_judge_verify_helper-4.10.1-py3.8.egg\onlinejudge_verify\marker.py", line 37, in get_current_timestamp
    depending_files = language.list_dependencies(path, basedir=pathlib.Path.cwd())
  File "C:\Users\tsutaj\tmp\venv\lib\site-packages\online_judge_verify_helper-4.10.1-py3.8.egg\onlinejudge_verify\languages\cplusplus.py", line 168, in list_dependencies
    return _cplusplus_list_depending_files(path.resolve(), CXX=env.CXX, joined_CXXFLAGS=joined_CXXFLAGS)
  File "C:\Users\tsutaj\tmp\venv\lib\site-packages\online_judge_verify_helper-4.10.1-py3.8.egg\onlinejudge_verify\languages\cplusplus.py", line 47, in _cplusplus_list_depending_files
    data = subprocess.check_output(command)
  File "C:\Users\tsutaj\AppData\Local\Programs\Python\Python38\lib\subprocess.py", line 411, in check_output
    return run(*popenargs, stdout=PIPE, timeout=timeout, check=True,
  File "C:\Users\tsutaj\AppData\Local\Programs\Python\Python38\lib\subprocess.py", line 512, in run
    raise CalledProcessError(retcode, process.args,
subprocess.CalledProcessError: Command '['C:\\MinGW\\bin\\g++.EXE', '--std=c++17', '-O2', '-Wall', '-g', '-I', 'C:\\Users\\tsutaj\\tmp', '-MD', '-MF', '/dev/stdout', '-MM', 'C:\\Users\\tsutaj\\tmp\\examples\\debug\\gcc_only.test.cpp']' returned non-zero exit status 1.
WARNING:onlinejudge_verify.verify:failed to increase the stack size
::warning ::failed to ulimit
INFO:onlinejudge_verify.verify:verify: examples\monoids.hpp
ERROR:onlinejudge_verify.verify:PROBLEM is not specified
::error file=examples\monoids.hpp::failed to verify examples\monoids.hpp
ERROR:onlinejudge_verify.verify:1 tests failed
ERROR:onlinejudge_verify.verify:failed: examples\monoids.hpp

(venv) C:\Users\tsutaj\tmp>oj-verify run examples\nim

(venv) C:\Users\tsutaj\tmp>ls examples
awk           monoids.hpp       segment_tree.point_set_range_composite.test.cpp  union_find_tree.hpp
csharpscript  nim               segment_tree.range_minimum_query.test.cpp        union_find_tree.yosupo.test.cpp
debug         segment_tree.hpp  segment_tree.range_sum_query.test.cpp
macros.hpp    segment_tree.md   union_find_tree.aoj.test.cpp

(venv) C:\Users\tsutaj\tmp>oj-verify run example.test.cpp
no module named resource (skipped)
INFO:onlinejudge_verify.config:config file loaded: .verify-helper\config.toml: {'languages': {'awk': {'compile': "bash -c 'echo hello > {tempdir}/hello'", 'execute': 'env AWKPATH={basedir} awk -f {path}', 'bundle': 'false', 'list_dependencies': 'sed \'s/^@include "\\(.*\\)"$/\\1/ ; t ; d\' {path}'}}}
WARNING:onlinejudge_verify.languages:config.toml: languages.awk: Adding new languages using `config.toml` is supported but not recommended. Please consider making pull requests for your languages, see https://github.com/kmyk/online-judge-verify-helper/issues/116
<built-in>: fatal error: opening dependency file /dev/stdout: No such file or directory
compilation terminated.
Traceback (most recent call last):
  File "C:\Users\tsutaj\tmp\venv\lib\site-packages\online_judge_verify_helper-4.10.1-py3.8.egg\onlinejudge_verify\marker.py", line 37, in get_current_timestamp
    depending_files = language.list_dependencies(path, basedir=pathlib.Path.cwd())
  File "C:\Users\tsutaj\tmp\venv\lib\site-packages\online_judge_verify_helper-4.10.1-py3.8.egg\onlinejudge_verify\languages\cplusplus.py", line 168, in list_dependencies
    return _cplusplus_list_depending_files(path.resolve(), CXX=env.CXX, joined_CXXFLAGS=joined_CXXFLAGS)
  File "C:\Users\tsutaj\tmp\venv\lib\site-packages\online_judge_verify_helper-4.10.1-py3.8.egg\onlinejudge_verify\languages\cplusplus.py", line 47, in _cplusplus_list_depending_files
    data = subprocess.check_output(command)
  File "C:\Users\tsutaj\AppData\Local\Programs\Python\Python38\lib\subprocess.py", line 411, in check_output
    return run(*popenargs, stdout=PIPE, timeout=timeout, check=True,
  File "C:\Users\tsutaj\AppData\Local\Programs\Python\Python38\lib\subprocess.py", line 512, in run
    raise CalledProcessError(retcode, process.args,
subprocess.CalledProcessError: Command '['C:\\MinGW\\bin\\g++.EXE', '--std=c++17', '-O2', '-Wall', '-g', '-I', 'C:\\Users\\tsutaj\\tmp', '-MD', '-MF', '/dev/stdout', '-MM', 'C:\\Users\\tsutaj\\tmp\\example.test.cpp']' returned non-zero exit status 1.
<built-in>: fatal error: opening dependency file /dev/stdout: No such file or directory
compilation terminated.
Traceback (most recent call last):
  File "C:\Users\tsutaj\tmp\venv\lib\site-packages\online_judge_verify_helper-4.10.1-py3.8.egg\onlinejudge_verify\marker.py", line 37, in get_current_timestamp
    depending_files = language.list_dependencies(path, basedir=pathlib.Path.cwd())
  File "C:\Users\tsutaj\tmp\venv\lib\site-packages\online_judge_verify_helper-4.10.1-py3.8.egg\onlinejudge_verify\languages\cplusplus.py", line 168, in list_dependencies
    return _cplusplus_list_depending_files(path.resolve(), CXX=env.CXX, joined_CXXFLAGS=joined_CXXFLAGS)
  File "C:\Users\tsutaj\tmp\venv\lib\site-packages\online_judge_verify_helper-4.10.1-py3.8.egg\onlinejudge_verify\languages\cplusplus.py", line 47, in _cplusplus_list_depending_files
    data = subprocess.check_output(command)
  File "C:\Users\tsutaj\AppData\Local\Programs\Python\Python38\lib\subprocess.py", line 411, in check_output
    return run(*popenargs, stdout=PIPE, timeout=timeout, check=True,
  File "C:\Users\tsutaj\AppData\Local\Programs\Python\Python38\lib\subprocess.py", line 512, in run
    raise CalledProcessError(retcode, process.args,
subprocess.CalledProcessError: Command '['C:\\MinGW\\bin\\g++.EXE', '--std=c++17', '-O2', '-Wall', '-g', '-I', 'C:\\Users\\tsutaj\\tmp', '-MD', '-MF', '/dev/stdout', '-MM', 'C:\\Users\\tsutaj\\tmp\\examples\\debug\\gcc_only.test.cpp']' returned non-zero exit status 1.
WARNING:onlinejudge_verify.verify:failed to increase the stack size
::warning ::failed to ulimit
INFO:onlinejudge_verify.verify:verify: example.test.cpp
INFO:onlinejudge._implementation.logging:[x] problem recognized: AOJProblem.from_url('http://judge.u-aizu.ac.jp/onlinejudge/description.jsp?id=ALDS1_4_B'): https://onlinejudge.u-aizu.ac.jp/courses/lesson/1/ALDS1/4/ALDS1_4_B
INFO:onlinejudge_verify.verify:problem: https://onlinejudge.u-aizu.ac.jp/courses/lesson/1/ALDS1/4/ALDS1_4_B
INFO:onlinejudge_verify.verify:$ sleep 2
INFO:onlinejudge_verify.verify:$ oj download --system -d '.verify-helper\cache\42a547747c50ff5e0fe5f5448f36ffcc\test' --silent https://onlinejudge.u-aizu.ac.jp/courses/lesson/1/ALDS1/4/ALDS1_4_B
[ERROR] Failed to download since file already exists: '.verify-helper\cache\42a547747c50ff5e0fe5f5448f36ffcc\test'\testcase_00.in
Traceback (most recent call last):
  File "C:\Users\tsutaj\tmp\venv\lib\site-packages\online_judge_verify_helper-4.10.1-py3.8.egg\onlinejudge_verify\verify.py", line 92, in verify_file
    exec_command(command)
  File "C:\Users\tsutaj\tmp\venv\lib\site-packages\online_judge_verify_helper-4.10.1-py3.8.egg\onlinejudge_verify\verify.py", line 49, in exec_command
    subprocess.check_call(command)
  File "C:\Users\tsutaj\AppData\Local\Programs\Python\Python38\lib\subprocess.py", line 364, in check_call
    raise CalledProcessError(retcode, cmd)
subprocess.CalledProcessError: Command '['oj', 'download', '--system', '-d', "'.verify-helper\\cache\\42a547747c50ff5e0fe5f5448f36ffcc\\test'", '--silent', 'https://onlinejudge.u-aizu.ac.jp/courses/lesson/1/ALDS1/4/ALDS1_4_B']' returned non-zero exit status 1.
::error file=example.test.cpp::failed to verify example.test.cpp
ERROR:onlinejudge_verify.verify:1 tests failed
ERROR:onlinejudge_verify.verify:failed: example.test.cpp

(venv) C:\Users\tsutaj\tmp>echo %ERRORLEVEL%
1
```

Windows でこれと似た概念は `CON` で、例えば  `echo abc > /dev/stdout` はだめですが `echo abc > CON` は動いたりします。なのでこれらを置換すれば動く可能性はあります

tsutaj commented:
手元で `CON` に置換するとその手のエラーは出なくなったような気がしますが、以下もなんかおかしいですね・・・

```
[ERROR] Failed to download since file already exists: '.verify-helper\cache\42a547747c50ff5e0fe5f5448f36ffcc\test'\testcase_00.in
```

`oj download` が `shlex.quote` 周りでバグっているような予感がします

kmyk commented:
:scream:

kmyk commented:
`verify.py` の次の行の `shlex.quote` が不要ですね。Windows 以外だと quote が発生しないのでバグが発見されないままになってたみたいです。

https://github.com/online-judge-tools/verification-helper/blob/3df420fd40019fce4e296e116090a8f8deab6b83/onlinejudge_verify/verify.py#L83

tsutaj commented:
ありがとうございます、そこに関してはそれで解決ですね。

依存関係を得るところでバグるみたいです。(以下には一部、自分のデバッグ出力が含まれます)

```
[x] slowest: 0.252139 sec  (for testcase_09)
[+] test success: 10 cases
--std=c++17 -O2 -Wall -g -I 'C:\Users\tsutaj\tmp'
['C:\\MinGW\\bin\\g++.EXE', '--std=c++17', '-O2', '-Wall', '-g', '-I', 'C:\\Users\\tsutaj\\tmp', '-MD', '-MF', 'CON', '-MM', 'C:\\Users\\tsutaj\\tmp\\example.test.cpp']
example.test.o: C:\Users\tsutaj\tmp\example.test.cpp
data = b'' # data の中身を見ると空
[] # それゆえ依存関係リストが空
Traceback (most recent call last):
  File "C:\Users\tsutaj\tmp\venv\Scripts\oj-verify-script.py", line 11, in <module>
    load_entry_point('online-judge-verify-helper==4.10.1', 'console_scripts', 'oj-verify')()
  File "C:\Users\tsutaj\tmp\venv\lib\site-packages\online_judge_verify_helper-4.10.1-py3.8.egg\onlinejudge_verify\main.py", line 216, in main
  File "C:\Users\tsutaj\tmp\venv\lib\site-packages\online_judge_verify_helper-4.10.1-py3.8.egg\onlinejudge_verify\main.py", line 76, in subcommand_run
  File "C:\Users\tsutaj\tmp\venv\lib\site-packages\online_judge_verify_helper-4.10.1-py3.8.egg\onlinejudge_verify\verify.py", line 152, in main
  File "C:\Users\tsutaj\tmp\venv\lib\site-packages\online_judge_verify_helper-4.10.1-py3.8.egg\onlinejudge_verify\marker.py", line 53, in mark_verified
  File "C:\Users\tsutaj\tmp\venv\lib\site-packages\online_judge_verify_helper-4.10.1-py3.8.egg\onlinejudge_verify\marker.py", line 43, in get_current_timestamp
ValueError: max() arg is an empty sequence
```

問題の箇所を手元でインタラクティブに実行するとこのような感じになります。

```
>>> subprocess.check_output(['C:\\MinGW\\bin\\g++.EXE', '--std=c++17', '-O2', '-Wall', '-g', '-I', 'C:\\Users\\tsutaj\\tmp', '-MD', '-MF', 'CON', '-MM', 'C:\\Users\\tsutaj\\tmp\\example.test.cpp'])
example.test.o: C:\Users\tsutaj\tmp\example.test.cpp
b''
```
単に `/dev/stdout` を `CON` に変えるだけだと、実行はされますがなぜか返り値に何もかえってこず、失敗します。

たぶんこれは Windows 特有の問題な気がするのですが、どうすれば解決するのかまだわかっていません・・・

kmyk commented:
だとすると、`CON` が対応するのは正確には `/dev/stdout` ではなく `/dev/tty` なのでしょう。

`-o` に指定するのはファイルです。ファイルを作ったり消したりは面倒なので「書き込んだ内容を標準出力に転送してくれる魔法のファイル」である `/dev/stdout` を使っていましたが、難しいようなので正攻法でいきましょう。
[tempfile.NamedTemporaryFile](https://docs.python.org/ja/3/library/tempfile.html#tempfile.TemporaryDirectory) を使って一時ディレクトリを作り、その中の一度本物のファイルに書き込んでから、それを読み出すのがよいはずです。

tsutaj commented:
このように実装してみたんですが、エラー出ました。
```
    with tempfile.NamedTemporaryFile(mode='w') as temp_file:
        command = [str(CXX), *shlex.split(joined_CXXFLAGS), '-MD', '-MF', temp_file.name, '-MM', str(path)]
        subprocess.check_output(command)
        temp_file.seek(0)
        data = temp_file.read()
        makefile_rule = shlex.split(data.decode().replace('\\\n', ''))
        return [pathlib.Path(path).resolve() for path in makefile_rule[1:]]
```

```
<built-in>: fatal error: opening dependency file C:\Users\tsutaj\AppData\Local\Temp\tmplv1mchvq: Permission denied
compilation terminated.
Traceback (most recent call last):
  File "C:\Users\tsutaj\tmp\venv\lib\site-packages\online_judge_verify_helper-4.10.1-py3.8.egg\onlinejudge_verify\marker.py", line 37, in get_current_timestamp
    depending_files = language.list_dependencies(path, basedir=pathlib.Path.cwd())
  File "C:\Users\tsutaj\tmp\venv\lib\site-packages\online_judge_verify_helper-4.10.1-py3.8.egg\onlinejudge_verify\languages\cplusplus.py", line 173, in list_dependencies
    return _cplusplus_list_depending_files(path.resolve(), CXX=env.CXX, joined_CXXFLAGS=joined_CXXFLAGS)
  File "C:\Users\tsutaj\tmp\venv\lib\site-packages\online_judge_verify_helper-4.10.1-py3.8.egg\onlinejudge_verify\languages\cplusplus.py", line 49, in _cplusplus_list_depending_files
    subprocess.check_output(command)
  File "C:\Users\tsutaj\AppData\Local\Programs\Python\Python38\lib\subprocess.py", line 411, in check_output
    return run(*popenargs, stdout=PIPE, timeout=timeout, check=True,
  File "C:\Users\tsutaj\AppData\Local\Programs\Python\Python38\lib\subprocess.py", line 512, in run
    raise CalledProcessError(retcode, process.args,
subprocess.CalledProcessError: Command '['C:\\MinGW\\bin\\g++.EXE', '--std=c++17', '-O2', '-Wall', '-g', '-I', 'C:\\Users\\tsutaj\\tmp', '-MD', '-MF', 'C:\\Users\\tsutaj\\AppData\\Local\\Temp\\tmplv1mchvq', '-MM', 'C:\\Users\\tsutaj\\tmp\\example.test.cpp']' returned non-zero exit status 1.
```
Permmission denied だそうです (AppData にファイルができているのは、何か関係があるのでしょうか、、、)

kmyk commented:
えー、ごめんなさい :bow:

[`tempfile.NamedTemporaryFile`](https://docs.python.org/ja/3/library/tempfile.html#tempfile.NamedTemporaryFile) のドキュメントには以下のようにあって、これでエラーになっているように見えます。

> 名前付き一時ファイルがまだ開かれている間にこの名前を使って再度ファイルを開くことができるかどうかは、プラットフォームによって異なります (Unix 上では可能ですが、 Windows NT 以降ではできません)。

`tempfile.NamedTemporaryFile` でも適切にやれば回避できますが、難しいので `tempfile.TemporaryDirectory` を使ってください。(この問題には途中で気付いてリンクや文章は修正したんだけど、一部そのままになってました)

kmyk commented:
ついでに先回りですが、[pathlib](https://docs.python.org/ja/3/library/pathlib.html#basic-use) を使ってください。`tempfile.TemporaryDirectory` を使うとパスの連結をしないとだめなのですが、これを `tempdir + "/a.out"` ってやると壊れます。Windows 上では区切りに `\` を使うためです。

tsutaj commented:
`_cplusplus_list_depending_files` 関数内の

```
makefile_rule = shlex.split(data.decode().replace('\\\n', ''))
```

で、`'\\\n'` を `''` に置換したい理由は何でしょうか？おそらく改行を無視したいとのことだとは思うのですが、これがあるためにパスのスラッシュ `\` が無くなって  Windows だとバグるようです。

tsutaj commented:
ごめんなさい、`replace` 消しても同様のバグがでたのでどちらかというと `shlex.split` がやばいみたいです

tsutaj commented:
`platform.uname().system == 'Windows` かどうかで実行環境を確認し、Windows かどうかの bool 値 を`is_windows` として変数に残し、

```
shlex.split(data.decode().replace('\\\n', ''), posix=not(is_windows)
```
とすると解決はしました。とりあえずはこれで実装してみますが、だめなケースなどあればまた教えてほしいです。

tsutaj commented:
- [ ] `resource` を使う箇所を、環境によって分岐もしくは `try ... except` で処理させる (処理は一か所にまとめる)
- [ ] `/dev/stdout` を使っている箇所を `TemporaryDirectory` の中にファイル出力させるよう変更する
- [ ] `verify.py` において不要な `shlex.quote` を消す
- [ ] `shlex.split` の option をどうつけるか環境によって分岐させる

tsutaj commented:
なぜか Mac で落ちますね (今回のとは別の問題でしょうか？)

```
clang: error: unknown argument: '-fpreprocessed'
```

が出ています。#254 でも同様の理由で落ちている気がして、元からあった問題のようです

tsutaj commented:
Mac はあまり詳しくなく、手元にないのでテストもできないのですが、これって g++ のふりをした clang が原因なんでしょうか？

```
 test_no_newline (tests.test_bundle.TestStringMethods) ... clang: error: unknown argument: '-fpreprocessed'
ERROR

======================================================================
ERROR: test_no_newline (tests.test_bundle.TestStringMethods)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/runner/runners/2.169.1/work/verification-helper/verification-helper/tests/test_bundle.py", line 16, in test_no_newline
    bundler.update(path)
  File "/Users/runner/runners/2.169.1/work/verification-helper/verification-helper/onlinejudge_verify/languages/cplusplus_bundle.py", line 191, in update
    uncommented_lines = get_uncommented_code(path, iquotes=self.iquotes, compiler=self.compiler).splitlines(keepends=True)
  File "/Users/runner/runners/2.169.1/work/verification-helper/verification-helper/onlinejudge_verify/languages/cplusplus_bundle.py", line 105, in get_uncommented_code
    code = _get_uncommented_code(path.resolve(), iquotes_options=iquotes_options, compiler=compiler)
  File "/Users/runner/runners/2.169.1/work/verification-helper/verification-helper/onlinejudge_verify/languages/cplusplus_bundle.py", line 100, in _get_uncommented_code
    return subprocess.check_output(command, shell=True)
  File "/Users/runner/hostedtoolcache/Python/3.8.2/x64/lib/python3.8/subprocess.py", line 411, in check_output
    return run(*popenargs, stdout=PIPE, timeout=timeout, check=True,
  File "/Users/runner/hostedtoolcache/Python/3.8.2/x64/lib/python3.8/subprocess.py", line 512, in run
    raise CalledProcessError(retcode, process.args,
subprocess.CalledProcessError: Command 'g++ -I /private/var/folders/24/8k48jl6d249_n_qfxwsl6xvm0000gn/T/tmprnx6hnhx -fpreprocessed -dD -E /private/var/folders/24/8k48jl6d249_n_qfxwsl6xvm0000gn/T/tmprnx6hnhx/example.test.cpp' returned non-zero exit status 1.

----------------------------------------------------------------------
```

kmyk commented:
> これって g++ のふりをした clang が原因なんでしょうか？

はい、おそらくそうです。こまった

kmyk commented on .github/workflows/test.yml:
```diff
@@ -4,7 +4,11 @@ on: [push, pull_request]
-    runs-on: ubuntu-latest
+    strategy:
+      matrix:
+        os: [ubuntu-latest, windows-latest, macos-latest]
```
```suggestion
        # os: [ubuntu-latest, windows-latest, macos-latest]  # now macos-latest doesn't work due to their bug, which g++ is not g++. see https://github.com/online-judge-tools/verification-helper/pull/255
        os: [ubuntu-latest, windows-latest]
```

とりあえず macOS を無効にします。
「パス区切りが `/` 以外だと動かない」は我々に責任のあるバグだけど「g++ が g++ 以外だと動かない」は我々に責任はない。矛盾した OS を出荷してる企業か、そういう変な環境を使ってるユーザが頑張るべき

kmyk commented:
私も mac は持っていないので、GitHub Actions 上では macOS を無効にすることで解決したい。(`Commit suggestion` のボタンを押してください。修正が取り込まれます)

kmyk commented:
Windows 上での CI が落ちてます。
一見 `oj-verify -h` が落ちてるように見えますが、他のと比較してよく見ると `python3 setup.py test` が悪そう。しかしもし `python3 setup.py test` が悪いとしても、さすがに無言で exit code 1 で落ちるのはなぜなのかよく分かりません。

``` console
▼Run oj-verify -h
  oj-verify -h
  python3 setup.py test
  shell: C:\Program Files\PowerShell\7\pwsh.EXE -command ". '{0}'"
  env:
    pythonLocation: C:\hostedtoolcache\windows\Python\3.8.2\x64
usage: oj-verify [-h] [--config-file CONFIG_FILE] {all,run,docs} ...

positional arguments:
  {all,run,docs}

optional arguments:
  -h, --help            show this help message and exit
  --config-file CONFIG_FILE
                        default: ".verify-helper/config.toml"
##[error]Process completed with exit code 1.
```

---

それはそうとして `tests/test_verify.py` が Library Checker を使っていて、いまの問題を解決すると次はそれで落ちそう。以下を http://judge.u-aizu.ac.jp/onlinejudge/description.jsp?id=ITP1_1_A などで書き換える必要があります。

https://github.com/online-judge-tools/verification-helper/blob/84a14c06fad5cf179a012dc15a4f9b8963f63a5f/tests/test_verify.py#L12-L30

kmyk commented:
しんどすぎないか？
@wakuwinmail たすけて

wakuwinmail commented:
流れを見たんですが、何が求められていますか

kmyk commented:
@wakuwinmail 雑すぎた。ちゃんと書きます。

お願いしたいことは以下のふたつです:

1. 発生している問題の原因を調べて @tsutaj に伝える
    - たとえば現状は https://github.com/online-judge-tools/verification-helper/pull/255#issuecomment-626333346 で困ってます
    - 放っておいても @tsutaj なら解決してくれるんですが、無限にコーナーケースが湧いてくるのしんどいはずなので手伝ってあげてほしい
1. このプルリクが完成したあと、実際の環境で使ってみて動くことを確認する

また、調査のためにこのプルリクのコードを使った `oj-verify` をインストールするには、以下のコマンドを実行してください。

``` console 
$ git clone https://github.com/online-judge-tools/verification-helper.git
$ cd verification-helper
$ git fetch origin pull/255/head
$ git checkout FETCH_HEAD
$ pip3 install -e .
```

wakuwinmail commented:
@tsutaj 
手元で試したところ、verify.pyの112行目から呼ばれたexec_command内44行目でエラーが発生しています。エラーメッセージ見たところ存在しないコマンドをsubprocess.check_call(command)に渡しているらしいです。
`Traceback (most recent call last):
  File "c:\verification-helper\onlinejudge_verify\verify.py", line 112, in verify_file
    exec_command(command)
  File "c:\verification-helper\onlinejudge_verify\verify.py", line 44, in exec_command
    subprocess.check_call(command)
  File "c:\python\python36-32\lib\subprocess.py", line 291, in check_call
    raise CalledProcessError(retcode, cmd)`

wakuwinmail commented:
subprocess.check_callの引数をコマンドプロンプトで実行すると[ERROR] no cases foundに、PowerShellで実行すると問題なく実行できるようです

kmyk commented:
なんだか `oj` コマンド側にも問題ありそうな気がする

@wakuwinmail 「subprocess.check_callの引数」が何なのか具体的に教えてほしいです

wakuwinmail commented:
@kmyk 
実行したのはこれです
`oj test -c .verify-helper\cache\2b68535019e98e921290d472204cd810\a.out -d '.verify-helper\cache\2b68535019e98e921290d472204cd810\test' --print-input --tle 60`

kmyk commented:
https://github.com/online-judge-tools/verification-helper/blob/5ae4fc9229d345075412ad17fa0a228d403b8f78/onlinejudge_verify/verify.py#L104

ここの `shlex.quote` が不要なだけぽい雰囲気ありますね

wakuwinmail commented:
shlex.quoteを消したらエラーが出なくなりました

tsutaj commented:
お二方ご報告ありがとうございます、修正しておきます

tsutaj commented:
> それはそうとして tests/test_verify.py が Library Checker を使っていて、いまの問題を解決すると次はそれで落ちそう。以下を http://judge.u-aizu.ac.jp/onlinejudge/description.jsp?id=ITP1_1_A などで書き換える必要があります。

これやったほうが良いですよね？書き換えます

tsutaj commented:
`cplusplus_bundle.py` のいくつかも書き換えが必要であるように思えます。手元で `python setup.py test` をした結果 bundle で落ちたからです。

具体的には以下が怪しいと思ってます (`shlex.quote` をなくせば動きます)

https://github.com/online-judge-tools/verification-helper/blob/84a14c06fad5cf179a012dc15a4f9b8963f63a5f/onlinejudge_verify/languages/cplusplus_bundle.py#L99

また、これ以外にも以下のエラーが出ていますが、原因調査中です。

```
======================================================================
ERROR: test_failure (tests.test_verify.TestStringMethods)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "c:\users\ybspx\tmp\verification-helper\tests\test_verify.py", line 79, in test_failure
    with onlinejudge_verify.marker.VerificationMarker(json_path=timestamps_path, use_git_timestamp=False) as marker:
  File "c:\users\ybspx\tmp\verification-helper\onlinejudge_verify\marker.py", line 28, in __init__
    self.load_timestamps(jobs=jobs)
  File "c:\users\ybspx\tmp\verification-helper\onlinejudge_verify\marker.py", line 93, in load_timestamps
    load(path, timestamp)
  File "c:\users\ybspx\tmp\verification-helper\onlinejudge_verify\marker.py", line 86, in load
    self.mark_failed(path)
  File "c:\users\ybspx\tmp\verification-helper\onlinejudge_verify\marker.py", line 64, in mark_failed
    path = path.resolve().relative_to(pathlib.Path.cwd())
  File "C:\Users\ybspx\AppData\Local\Programs\Python\Python38\lib\pathlib.py", line 899, in relative_to
    raise ValueError("{!r} does not start with {!r}"
ValueError: 'data\\example.test.cpp' does not start with 'C:\\Users\\ybspx\\AppData\\Local\\Temp\\tmpx9tt07cs'

======================================================================
FAIL: test_timestamps (tests.test_verify.TestStringMethods)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "c:\users\ybspx\tmp\verification-helper\tests\test_verify.py", line 117, in test_timestamps
    self.assertEqual(timestamps, expected_timestamps)
AssertionError: {'new-success.test.cpp': '2020-05-11 15:56:1[171 chars]900'} != {'not-updated.test.cpp': '2000-01-01 00:00:0[171 chars]900'}
  {'new-success.test.cpp': '2020-05-11 15:56:15 +0900',
-  'not-updated.test.cpp': '2020-05-09 23:05:13 +0900',
?                             ^   ^  ^^^^  ^ ^^

+  'not-updated.test.cpp': '2000-01-01 00:00:00 +0900',
?                             ^   ^  ^^^^  ^ ^^

   'updated-failure.test.cpp': '2000-01-01 00:00:00 +0900',
   'updated-success.test.cpp': '2020-05-11 15:56:15 +0900'}

----------------------------------------------------------------------
Ran 4 tests in 23.088s

FAILED (failures=1, errors=1)
Test failed: <unittest.runner.TextTestResult run=4 errors=1 failures=1>
error: Test failed: <unittest.runner.TextTestResult run=4 errors=1 failures=1>
```

tsutaj commented:
test で使用する問題は AOJ の http://judge.u-aizu.ac.jp/onlinejudge/description.jsp?id=ITP1_1_B に変えて (まだ commit はしていない) 、前の yosupo judge の問題コードは消してあります。方針に問題があれば直すので教えてください。

kmyk commented:
> `cplusplus_bundle.py` も書き換えが必要であるように思えます。手元で `python setup.py test` をした結果 bundle で落ちたからです。

その修正は正しそうです。お願いします。

> また、これ以外にも以下のエラーが出ていますが、原因調査中です。

ひとつめのエラーは次の行の `"data/example.test.cpp"` が原因に見えます。`str(pathlib.Path("data", "example.test.cpp"))` に変更してみてください。

https://github.com/online-judge-tools/verification-helper/blob/84a14c06fad5cf179a012dc15a4f9b8963f63a5f/tests/test_verify.py#L71

ふたつめのエラーは [`os.utime`](https://docs.python.org/ja/3/library/os.html#os.utime) に起因するはずです。
「更新がなかったファイルのタイムスタンプは変更されずそのままである」ことを確認するために、作成時刻が十分古い (2000-01-01) のファイルを用意し、これを用いています。Windows ではなにやらタイムスタンプを巻き戻せない ([python-Bugs-1057992 os.utime() doesn't work on WinME](https://mail.python.org/pipermail/python-bugs-list/2004-November/025966.html); 要確認) らしいので、そのようなファイルが作れず、エラーになっているようです。
`os.utime` を使わない形ですべて書き直すか、[`@unittest.skipif(platform.system() == 'Windows', "os.utime() fails on Windows")`](https://docs.python.org/ja/3/library/unittest.html#unittest.skipIf) などを書いてテストを無効化してください。

https://github.com/online-judge-tools/verification-helper/blob/84a14c06fad5cf179a012dc15a4f9b8963f63a5f/tests/test_verify.py#L48-L50

---

> test で使用する問題は AOJ の http://judge.u-aizu.ac.jp/onlinejudge/description.jsp?id=ITP1_1_B に変えて (まだ commit はしていない) 、前の yosupo judge の問題コードは消してあります。方針に問題があれば直すので教えてください。

この方針も適切そうです。その方針でよろしくお願いします。

tsutaj commented:
> ひとつめのエラーは次の行の "data/example.test.cpp" が原因に見えます。str(pathlib.Path("data", "example.test.cpp")) に変更してみてください。

これ本当に謎なんですが、Windows だとこう書いたとしても `path.resolve()` したときに絶対パスになってくれないようです。  
こちらでもいろいろ試してみてたんですが結局どう書けば解決するのかはよくわかりません・・・。具体的には、`ValueError: 'data\\example.test.cpp' does not start with 'C:\\Users\\ybspx\\AppData\\Local\\Temp\\tmpku8o642q'` と出てしまい、相対関係がわからないという具合で怒られます。(ちなみに、Cygwin だと例えば `/tmp/tmpfqrz7eoh/data/example.test.cpp` と `/tmp/tmpfqrz7eoh` との比較になるので、正しく実行されます)

timestamp については、修正が楽という理由から、テストを無効化する方針で行きます。

kmyk commented:
- `mark_failed` が、存在しないファイルのパスが渡ってきたときのことを想定してない
  https://github.com/online-judge-tools/verification-helper/blob/84a14c06fad5cf179a012dc15a4f9b8963f63a5f/onlinejudge_verify/marker.py#L63-L65
- このテスト自体は `"data/example.test.cpp"` なのが間違いで `"example.test.cpp"` が正しい (ファイルが存在してかつ verify 失敗した場合をテストしたいので)
- (削除されたファイルがあった場合のテストがない)

こういうことな気がしてきました

kmyk commented:
`mark_failed` はとりあえず `if path.exists():` ってやるとよさそう

tsutaj commented:
なるほど、ありがとうございます。最も楽そうな解決方法として、`str(pathlib.Path("example.test.cpp"))` としてしまう (`data/` をなくす) というのがあって、これで対処するとうまく動くのですが、これで大丈夫でしょうか？

`mark_failed` で path が存在するかどうかのチェックは入れておきます。

あと、timestamp に関連する話で、以下もテストが落ちてまずいようですが、これも Windows の場合は無視するように変えてしまって大丈夫ですか？

https://github.com/online-judge-tools/verification-helper/blob/3dbf0951ae4c01e0eda9eb599d9a19b88805ac19/tests/test_verify.py#L85

tsutaj commented:
上を全部反映させたら手元で `OK (skipped=1)` がでた

kmyk commented:
> str(pathlib.Path("example.test.cpp")) としてしまう (data/ をなくす) ので大丈夫でしょうか？

はい、大丈夫です。ただし、これは `str` も `pathlib.Path` も要らないので直接 `"example.test.cpp"` と書いてくれるとよりうれしいです

> あと、timestamp に関連する話で、以下もテストが落ちてまずいようですが、これも Windows の場合は無視するように変えてしまって大丈夫ですか？

assert してる部分を通るように書き換えるのがおすすめです。(テストの入力自体が修整されてるので出力も変わるはず)

> まずい変更方針じゃなかったら push します

よさそう。おねがいします
(もし実はまずかったとしても後から修整すればよいので、そちらがよさそうだと思ったら push してくれて大丈夫です)

tsutaj commented:
なぜか Ubuntu のほうが通らなくなって泣いてしまった (`self.assertEqual(timestamps, {})` のところで落ちているので、たぶん自分が変なことをしてしまっています)

tsutaj commented:
細かく追ってはいないのですが、どうも `str(pathlib.Path("data", "example.test.cpp"))` としないと Ubuntu 側がバグる (タイムスタンプがうまく合ってくれない？) ようです。

`mark_failed` のほうでパスの存在確認さえ行えば、このように書いてもテストは通るので `example.test.cpp` を元に戻します

tsutaj commented:
> ただし、これは str も pathlib.Path も要らないので直接 "example.test.cpp" と書いてくれるとよりうれしいです

すみませんこれ忘れてました (対応します)

tsutaj commented:
直しました

ところで Windows まだ落ちてる・・・ (どうして？？)

tsutaj commented:
@wakuwinmail 手元の Windows (cmd に python 入れた環境) で再現されるバグは全て取ったのですが、未だ CI で落ちる、という状況です。

ローカルリポジトリを最新の状態にして、実行したときにエラーがでるかどうか教えてもらえますか？その情報をもとにまたいろいろ対応したいと思います

wakuwinmail commented:
自分の環境でoj-verify runしたんですが、全テスト通過してますね
oj-verify -hも問題なく実行できました

tsutaj commented:
@wakuwinmail 報告ありがとう、そちらでも問題なく動くんですね・・・

エラーが出ないことには対処しようがないので、手元で他の動かし方を試してみることにします。
